src/extractKey.ts

```diff
@@ -0,0 +1,83 @@
+import { createOperation, Operation } from 'apollo-link';
+import { DirectiveNode, ArgumentNode } from 'graphql';
+
+import deepUpdate = require('deep-update');
+
+export function extractKey(operation: Operation): { operation: Operation, key?: string } {
+    // Explicit keys in the link context win out.
+    const { serializationKey } = operation.getContext();
+    if (serializationKey) {
+        return { operation, key: serializationKey };
+    }
+
+    const { directive, path } = extractDirective(operation);
+    if (!directive) {
+        return { operation };
+    }
+    const argument = directive.arguments.find(d => d.name.value === 'key');
+    if (!argument) {
+        throw new Error(`The @serialize directive requires a 'key' argument`);
+    }
+    let key = valueForArgument(argument, operation.variables);
+    // Replace any {{variable}}s with their value.
+    key = key.replace(/\{\{([^\}]+)\}\}/g, (_substring, name) => {
+        return getVariableOrDie(operation.variables, name);
+    });
+
+    // Pass through the operation, with the directive removed so that the server
+    // doesn't see it.
+    const finalIndex = path.pop();
+    const newOperation = createOperation(operation.getContext(), {
+        ...operation as any,
+        query: deepUpdate(operation.query, path, {$splice: [[finalIndex, 1]]}),
+    });
+
+    return { operation: newOperation, key };
+}
+
+function extractDirective({ query, operationName }: Operation): { directive?: DirectiveNode, path?: string[] } {
+    const path: string[] = [];
+
+    // First, find the operation definition
+    let operationNode;
+    for (let i = 0; i < query.definitions.length; i++) {
+        const node = query.definitions[i];
+        if (node.kind !== 'OperationDefinition') {
+            continue;
+        }
+        if (!operationName || node.name.value == operationName) {
+            operationNode = node;
```

Comment:
This assumes that there's only one operation definition, or at the very least that the serialization directive is on the first operation definition. That's probably fine to avoid extra complexity, but I think it's worth documenting